lib/rtti.h

```diff
@@ -0,0 +1,313 @@
+
+    [[nodiscard]] static constexpr TypeId id() noexcept {
+        return detail::TypeIdResolver<T>::resolve();
+    }
+
+    [[nodiscard]] static constexpr bool isA(TypeId typeId) noexcept {
+        return (id() == typeId) || (... || (Parents::TypeInfo::isA(typeId)));
+    }
+
+    template <typename T>
+    [[nodiscard]] static const void *dyn_cast(TypeId typeId, const T *ptr) noexcept {
+        if (id() == typeId) return static_cast<const This *>(ptr);
+
+        std::array<const void *, sizeof...(Parents)> ptrs = {
+            Parents::TypeInfo::dyn_cast(typeId, ptr)...};
+
+        auto it =
+            std::find_if(ptrs.begin(), ptrs.end(), [](const void *ptr) { return ptr != nullptr; });
+        return (it != ptrs.end()) ? *it : nullptr;
+    }
+};
+
+struct Base {
+    virtual ~Base() = default;
+
+    /// Queries information about dynamic type of an object.
+    /// @return typeid of current object.
+    [[nodiscard]] virtual TypeId typeId() const noexcept = 0;
+    /// Checks if given object is of specified dynamic type.
+    /// @param typeid to check with
+    [[nodiscard]] virtual bool isA(TypeId typeId) const noexcept = 0;
+
+    /// Checks if given object is of specified dynamic type.
+    /// Same as `isA`, but typeid to check with is derived from template argument.
+    template <typename T>
+    [[nodiscard]] bool is() const noexcept {
+        return isA(TypeInfo<T>::id());
+    }
+
+    /// Casts current object to a given type `T`.
+    /// Overall the semantics is similar to that of `dynamic_cast`, however,
+    /// relies on custom RTTI metadata and intrusive annotation.
+    /// @return nullptr if object is not compatible with `T` (e.g. if `T` is not
+    /// among its base classes), casted pointer to object of T* type otherwise.
+    template <typename T>
+    [[nodiscard]] T *to() noexcept {
+        return reinterpret_cast<T *>(const_cast<void *>(toImpl(TypeInfo<T>::id())));
+    }
+
+    /// Same as `to`, but returns const pointer to T.
+    template <typename T>
+    [[nodiscard]] const T *to() const noexcept {
+        return reinterpret_cast<const T *>(toImpl(TypeInfo<T>::id()));
+    }
+
+ protected:
+    [[nodiscard]] virtual const void *toImpl(TypeId typeId) const noexcept = 0;
+};
+
+}  // namespace RTTI
+
+/// Declare typeinfo for a given class `T`.
+///
+/// All direct bases (including virtual) should be specified as additional
+/// arguments. For example:
+/// struct Base1 {
+///   ...
+///   DECLARE_TYPEINFO(Base1);
+/// };
+/// struct Base2 {
+///   ...
+///   DECLARE_TYPEINFO(Base2);
+/// };
+///
+/// struct Derived : public Base1, public virtual Base2 {
+///   ...
+///   DECLARE_TYPEINFO(Derived, Base1, Base2);
+/// };
+#define DECLARE_TYPEINFO(T, ...)                                                  \
+ private:                                                                         \
+    static constexpr RTTI::TypeId static_typeId() { return RTTI::InvalidTypeId; } \
```

Comment:
Rather than doing all the complex hasing stuff, why not just something like

    static constexpr RTTI::TypeId static_typeId() { return reinterpret_cast<uintptr_t>(&static_typeId); }

That way, the linker will provide a distinct value (address) for each class automatically.

Reply:
This inhibits many optimizations for `is<>` or where typeid is involved. Having typeid as a compile-time constant (instead of link-time) allows for some nice simplifications in the set of comparisons.